### ISSUE TITLE
We should be able to develop and test Quiet without Tor

### ISSUE THREAD

**Issue by holmesworcester** [TryQuiet/quiet#2286]
Right now Quiet's p2p functionality depends on Tor. Tor is a great fallback because onion address don't change, and because we can assume Tor will continue to provide this relaying service going forward. It's also a good option for users who want some metadata and IP address privacy. (Or, in the future, for users on restrictive networks.) However, Tor has some problems:

1. Onion addresses take too long to become available. 
2. Onion addresses sometimes take too long to connect. 
3. Sometimes connection fails
4. As a result, e2e tests based on Tor are also unreliable
5. Also as a result, it slows down developers who need to test that something is working
6. Users experience unexpected delays
7. Tor is slower than direct connections, so Quiet is not as fast as it could be in situations where peers could connect directly (peers on local networks, peers with forwarded ports)  

What are some solutions? At a high level here are some:

1. **Direct peer connection.** Peers connect directly when they can, for users, and in tests and local development we make sure they can.
2. **Server assist.** We use the store and forward service in all situations where we can't wait for Tor, and we make sure to have some long-running e2e tests, with really long timeouts, to make sure everything still works with Tor as a fallback. 
3. **Mock Tor.** Here were create a mock Tor service (running on some central service) that lets us mess with Tor's parameters. This would let us make Tor unrealistically fast for testing purposes, and then make it brutally slow or high latency to make sure Quiet still works, when we want to do that.

For #2, we would want to move as quickly as possible to being able to use the store and forward service for joining and updating community state with the desktop app. This is pretty much the path we're on.

For #1, we could think of this in a very hacky way confined to tests or developer experience, i.e. just do whatever works to make two Quiets connect via TCP instead of Tor. This is probably easy but will have limitations and Quiet might not always work as expected. 

Alternatively, we could think of it as shipping a feature that will benefit users. For example, Quiet could optionally make directly TCP connections when peers add a recent TCP address and a free port to OrbitDB. Or peers could just add their latest IP address and port to OrbitDB and we could try it even if it's likely to fail in most cases. 

Or we could enable MDNS, for local libp2p peer discovery. See: https://github.com/libp2p/specs/blob/master/discovery/mdns.md (NOTE: of all the options, this one might be the easiest to implement.) Then Quiet would work blazingly fast for local development and for users on the same wifi netowrk. 

#3 is probably necessary for ensuring Quiet works well over Tor, because to ensure Quiet works well in all conditions we need to simulate Tor in all possible conditions and test to make sure Quiet works.

**Comment by kowalski** [comment]
I think "direct TCP connection" path is way too low level. Solving nat traversal is a big project.

We've tried in the past the webrtc transport of libp2p. It works with a signaling server. It's a tiny server that allows clients to make websocket connections. App peers connect to it. When peer wants to make a connection to a different peer it uses its PeerId to identify it. Signaling server exchanges the handshake data required to establish the webrtc connection. Once connection is estabilised the peers are connected directly. Signaling server does not receive participate in gossip or pubsub protocols.
Webrtc is the protocol that most online conference software uses. So for example Google meets uses it to send audio and video.
Since it's implemented in libp2p we would not have to change the higher levels of the app. Simply adding Signaling server and configuration of the transport should make libp2p use it to dial to other peers.

**Comment by holmesworcester** [comment]
One note is that because of how hole punching works, I think we need a TURN service that relays all data in order to support two clients behind the same NAT, for example a developer testing Quiet. mDNS might work better in this case even if it's less generally useful. Also running our own TURN service could be fine. We might want to restrict it to registered users at some point.

**Comment by kowalski** [comment]
I remember we were using libp2p webrtc transport from the office on multiple devices and there was no issue

**Comment by holmesworcester** [comment]
Looks like some routers support it and some don't: https://en.wikipedia.org/wiki/Network_address_translation#NAT_hairpinning

There's also UPnP port forwarding which, if enabled, makes holepunching not necessary.

I guess this means a developer is likely to be able to get it working one way or another, and for users it will work in 10-20% of cases without the TURN server, according to stats I've seen. So yes this is a good option!

**Comment by kowalski** [comment]
I guess my point was, that we should find a high-level tool that allows us not to become PhD students on network routing and low level protocols.
So I was looking for something off-the-shelf

Tests we've made with libp2p webrtc transport were definitely working in more than 20% cases

**Comment by holmesworcester** [comment]
Agree! And I meant to say that it fails in 10-20 percent of cases between peers in the wild. So it works in 80-90 percent of cases. But if you're all behind friendly routers it will maybe never fail!

**Comment by holmesworcester** [comment]
Libp2p supports webrtc without STUN servers: 
https://docs.libp2p.io/guides/getting-started/webrtc

Other overlay networks:
* https://py-ipv8.readthedocs.io/en/latest/
* https://www.pinecone.io/
